bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java

```diff
@@ -219,201 +288,432 @@ protected void updatedModuleHandlerFactory(ModuleHandlerFactory moduleHandlerFac
      */
-    Rule initRuleId(Rule rule) {
+    protected void updateRule(Rule rule, boolean enabled) {
         String rUID = rule.getUID();
-        Rule ruleWithUID;
-        if (rUID == null) {
-            synchronized (this) {
-                rUID = getRuleUID(rUID);
-                ruleWithUID = new Rule(rUID, rule.getTriggers(), rule.getConditions(), rule.getActions(),
-                        rule.getConfigurationDescriptions(), rule.getConfiguration(), rule.getTemplateUID(),
-                        rule.getVisibility());
-                ruleWithUID.setName(rule.getName());
-                ruleWithUID.setTags(rule.getTags());
-                ruleWithUID.setDescription(rule.getDescription());
-            }
+        if (getRuntimeRule(rUID) == null) {
+            logger.debug("There is no rule with UID '{}' which could be updated", rUID);
+            return;
+        }
+        RuntimeRule runtimeRule = new RuntimeRule(rule);
+        if (enabled) {
+            setRuleStatusInfo(rUID, new RuleStatusInfo(RuleStatus.NOT_INITIALIZED), runtimeRule);
+            setRule(runtimeRule);
         } else {
-            ruleWithUID = rule;
+            setRuleStatusInfo(rUID, new RuleStatusInfo(RuleStatus.DISABLED), runtimeRule);
         }
-        return ruleWithUID;
+        logger.debug("Rule with UID '{}' is updated.", rUID);
     }
 
     /**
-     * Validates ids of modules. The module id must not contain dot.
+     * This method runs a {@link Rule}. It is called by the {@link RuleEngineCallback}'s thread when a new
+     * {@link TriggerData} is available. This method switches
      *
-     * @param modules list of trigger, condition and action modules
-     * @throws IllegalArgumentException when a module id contains dot.
+     * @param rule the {@link Rule} which has to evaluate new {@link TriggerData}.
+     * @param td {@link TriggerData} object containing new values for {@link Trigger}'s {@link Output}s
      */
-    protected void validateModules(List<Module> modules) {
-        for (Module m : modules) {
-            String mId = m.getId();
-            if (mId == null || !mId.matches("[A-Za-z0-9_-]*")) {
-                throw new IllegalArgumentException("Invalid module uid: " + mId != null ? mId
-                        : "null" + ". It must not be null or not fit to the pattern: [A-Za-z0-9_-]*");
+    protected void runRule(RuntimeRule rule, RuleEngineCallbackImpl.TriggerData td) {
+        String rUID = rule.getUID();
+        RuleStatusInfo running = new RuleStatusInfo(RuleStatus.RUNNING);
+        synchronized (this) {
+            RuleStatus ruleStatus = getRuleStatus(rUID);
+            if (ruleStatus == RuleStatus.DISABLED || ruleStatus == RuleStatus.NOT_INITIALIZED) {
+                logger.error("Trying to execute rule ‘{}' with status '{}'", rUID, ruleStatus.getValue());
+                return;
+            }
+            if (ruleStatus == RuleStatus.RUNNING) {
```

Comment:
Too complex logic. The rule has to be started only when it is in IDLE state and it must go back from RUNNING to IDLE when it finished its work. If it is disabled in the middle of rule execution (when it is in RUNNING state) It has to stop working and go into DISABLED state.